Graphql Error Handling

kb-0311 opened the pull request:
<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**

<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Issue Number:**

Fixes #567

**Did you add tests for your changes?**

<!--Yes or No. Note: Add unit tests or automation tests for your code.-->

**Snapshots/Videos:**

<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**

<!--Add link to Talawa-Docs.-->

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-docs/blob/develop/CONTRIBUTING.md)?**

<!--Yes or No-->

kb-0311 committed:
gql errors

kb-0311 committed:
typo

kb-0311 commented:
This doc was based on the conversation I had with @xoldyckk where a lot of the ideas behind approach was cleared .

xoldd commented:
@kb-0311 You forgot to provide [this link](https://productionreadygraphql.com/2020-08-01-guide-to-graphql-errors/).

The doc is not up to mark. You should mention how this approach is different from just throwing graphql errors. The difference is that errors here are a part of the schema, so they're type safe and predictable, instead of root errors list which can return just about anything. Don't just copy paste the stuff from discussions, write them in a clean way with proper structure and proper graphql SDL and typescript resolver example.

Make seperate sections on how errors are to be handled. The sections should be for query(or mutation) resolvers, field resolvers for complex object fields and field resolvers for scalar fields. Use sensible, real world examples for errors to make it easier to understand. Btw that `postCount` example of yours inside an error object is wrong, it's not how partial object fields are going to be handled. `postCount` doesn't need to be nested inside the connection object, it can very well exist as a seperate field inside the `User` type.

Explain why some object fields are not nested directly and instead are making use of union types, how partial errors with data are going to be returned, and how clients will need to handle them.

kb-0311 commented:
@xoldyckk Yeah this is still a work in progress. I will refactor the doc with what you suggested including the graphql in production link and the basis of that approach in more details.

>Btw that postCount example of yours inside an error object is wrong, it's not how partial object fields are going to be handled. postCount doesn't need to be nested inside the connection object, it can very well exist as a separate field inside the User type.

No, it is correct as it already is. I had already thought about all the possible approaches to handling partial data in case of specific errors and how its implementation would look on the client side before reaching that approach in this comment. https://github.com/PalisadoesFoundation/talawa-api/discussions/1214#discussioncomment-5400166.

`postCount` is not a User attribute nor is `PrivatePostsError` related to User in any way, it is a Post attribute more appropriately. Remember, organizations can throw `PrivatePostsError` too for now.

Let's say I make the `postCount` field in the type `User` then I will have to put that field for Organization and in the future, (if something like sub-communities within an organization is implemented ) places where we can query private posts will need to have this `postCount` attribute too. Not only is that repetitive when it need not be, but in the client, if the posts are resolved then a simple `.length()` can set the state of the UI element, with no need for a count variable for the parent object.

By making sure that the data we want to send is with the Error object itself we have made sure that client knows what they are getting with the error check itself. It is not just a `PrivatePostsError` thing I mean almost all errors where we would want to send some kind of information related to the Error should be done this way for example `BlockedError` if we want to send `blockedByUser` info then it will need to be defined like that too, within the Error Object. This way certain Error types can also be extended and its attributes inherited by other Errors when needed.

Think about in terms of client apps and how the UI elements will work in terms of all other approaches and this is approach would make client-side behavior the simplest without any null checks within the resolved objects and stuff. I don't think any other approach would make this implementation in the react component simpler  https://github.com/PalisadoesFoundation/talawa-api/discussions/1214#discussioncomment-5424581. 

Check out Laurin's blog from the guild about Handling Errors,

https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/

xoldd commented:
@kb-0311 The thing you're saying about returning data inside errors is non-sensical. No where in any blog have I ever read this kind of approach. The blog you linked me doesn't mention anything about this approach either. All these blogs only mention defining errors for query and mutation fields. None of them have ever talked about implementing error in schema approach for nested fields. Because they assume that we should use a mix of their approach + throwing errors in nested resolvers.

What I want is error to be consolidated for both query/mutation resolvers and the nested fields too. Now which nested fields need that kind of consolidation depends on the access control logic of the field for different kinds of clients.

The thing about `postsCount` field not related to user is completely wrong. It is related to both user and organization, and if need be it needs to be defined as a field for both of those objects. The difference is that postsCount for a user is different from postsCount for an organization. `postsCount` for a user includes all the posts they've ever made in any organization. `postsCount` for an organization includes all the posts that are created by any number of members of that particular organization.

Counting posts manually on the client side is out of the question, that's never gonna happen. `postsCount` is provided by the server, it's not calculated on the client even if the number of posts is something as insignificant as the number `1`.

`postsCount` doesn't need to be resolved along with `postsConnection`, you can provide a `postsCount` field inside the connection too, but that's just for client's convenience. If `postsCount` should always be provided, but connection can be null, `postsCount` needs to exist as a seperate field. `postsCount` being a seperate field depends how `postsCount` is displayed on the client side. For example a posts button which shows number of posts, when you click on it, it makes a request for `postsConnection`, here's you don't need to resolve `postsConnection` just to get `postsCount`.

Think carefully and provide solid arguments that do actually create address a genuine problem. As far as I'm concerned your approach of returning `postsCount` along with an error message is as anti-graphql as it can be.

kb-0311 commented:
>The thing you're saying about returning data inside errors is non-sensical. No where in any blog have I ever read this kind of approach.

This is the blog linked by all the blogs on this topic.
https://sachee.medium.com/200-ok-error-handling-in-graphql-7ec869aec9bc

Go to the `Complex Schema Structure` subheading and take look at the `IsBlocked` Error. 

If the `blockedByUser` field mentioned there is not data returned within an error then I don't know what is.

Is this approach also "as anti-graphql as it can be" since that returns some form of data within the Error too which you are against? Now don;y nitpick what kind of data with error you were talking about in your comment.

The postCount one was just an example of sending data along with an error albeit a poorly chosen one.


>postsCount doesn't need to be resolved along with postsConnection, you can provide a postsCount field inside the connection too, but that's just for client's convenience. If postsCount should always be provided, but connection can be null, postsCount needs to exist as a seperate field. postsCount being a seperate field depends how postsCount is displayed on the client side. For example a posts button which shows number of posts, when you click on it, it makes a request for postsConnection, here's you don't need to resolve postsConnection just to get postsCount.

This is a good use case where something like my approach does not address. The problem with this approach is if when resolving the nested post connection we want to send an error that should not return a post count like "FlaggedPostsError "then what happens to the 'postCount' field?

Probably the only way out of this I could think of is making a `postCount` resolver which will resolve that field in the parent if we could need it other by making a db request for counting documents separately (User , Organization or anything ). In these kinds of cases then we will need to think about fields with these use cases.
 
My point about needing to return some form of resolved data within the Error still stands.

kb-0311 commented:
@xoldyckk Okay I spent my time going over this thing again and I basically answered all of my questions in my previous comment myself. You can successfully ignore that comment, now it is basically resolved jargon of nothing.

We will need some form of data resolved along with an error. Sending that within the error, which was my previous approach would lead to a lot of  headache when doing level wise and a lot of boilerplate we might not need, the more I went over its implementation details in my head. Plus we will need a lot of partial data in the future like during private events, querying group chats private event projects, roles etc and now after putting more thought into this I am not confident with how well my previous approach would scale. 

I went back to the drawing board again and then re-thought the thing about the 6a approach.

Remember how my main problem with the 6a approach in this comment was -
https://github.com/PalisadoesFoundation/talawa-api/discussions/1214#discussioncomment-5428485

>For other errors unspecific to fields, I thought it was maybe best not to send everything at once as an array since it could be a little annoying to get rid of all the notifications for errors as a mobile/web app user if we are listening for specific errors in the errors array in the payload. There could be ways around it but would require extra work on the client side.Like you know I am traversing the userErrors array in the eg and I am displaying a notification for each error in it yk .So I thought it would be best to have one error at a time when it came to API logic type errors to define the 5-second notification behavior. more accurately and send partial data with it.

But now I say that some of the query return types are nullable not only that some of the field within the return types are nullable too because of multiple reasons. That got me thinking.

```javascript
type CreateUserPayload {
  user: User
  userErrors: [CreateUserError!]!
}

union CreateUserError = UserNameTaken | PasswordTooShort | MustBeOver18

type UserNameTaken implements UserError {
  message: String!
  path: String!
  suggestion: String!
}

interface UserError {
  message: String!
  path: String!
}
```

>I thought it was maybe best not to send everything at once as an array since it could be a little annoying to get rid of all the notifications for errors as a mobile/web app user if we are listening for specific errors in the errors array in the payload.

1. We can make sure that the behavior I was saying could happen does not happen by returning from the resolver function at the appropriate time. That we can still have some form of atomicity for cases we really need. We need not push all errors into the userErrors array Array and return at the last minute.
2. Since Like I said some queries are nullable hell even some fields within the return type are nullable as well. In that case, we CAN send partial data along with the errors as two separate entities. 
3. Like for example in a resolver, I am returning the user as well as userErrors at the same time, the user may or may not be returned null or in certain errors within the userErrors we can return the user with only certain fields resolved and it a field or scalar throws errors send that field in user as null and put its corresponding error in the userErrors array.
4. If a field is null then that way a separate component could be mounted depending on its Error in the userError array.

@xoldyckk Do let me know what are your thoughts about this 6a approach.

kb-0311 commented:
@xoldyckk Pls do checkout this comment it goes over 6a approach in more detail and it should answer all of your points. 

https://github.com/PalisadoesFoundation/talawa-api/discussions/1239#discussioncomment-5486951

kb-0311 committed:
init

kb-0311 committed:
moved error docs to new location

kb-0311 committed:
fixed links

kb-0311 commented:
@xoldyckk Added more background information about the exact problem with the default approach of graphql errors with a brief introduction of thinking in terms of graph which will be expanded upon in a lot more detail in the api-design docs.

If this is good, I will be explaining in the documentation the 6a solution in terms of examples for all the cases we discussed in the GitHub discussions, which will be really long so made a separate doc for Error Handling.

kb-0311 committed:
typo

kb-0311 committed:
the practical cases

kb-0311 committed:
field level errors

kb-0311 commented:
@xoldyckk do tell me if you want to add more information.

Working on atomic errors and  nested resolver errors.

kb-0311 committed:
small refactor

kb-0311 committed:
atomic errors

kb-0311 committed:
nestes resolver errors

kb-0311 commented:
@xoldyckk Pls do review and tell me where all I need to expand more in detail.

kb-0311 committed:
small change

kb-0311 commented:
@palisadoes could you please re-request adarsh for a review?

xoldd commented on docs/developers/talawa-api/error-handling.md:
```diff
@@ -0,0 +1,625 @@
+
+Now the problems with this approach ->
+1. The `errors` key of a GraphQL response is not type safe. Graphql is loved because of the consistency and type safety it provides, we do not get the same with the `errors` array. They are specified, but introspection tells a client nothing about what might be found in them, they're harder to evolve, and harder to extend.
+2. By default if the `errors` array is present then the resulting query will return `null` in the `data` field. By that this means even if some nested resolver fails, that has the impact on the result of the root parent resolver (contradicts Introduction point 4).Essentially failure of a related node leads to the failure of parent node.This can potentially be a deal breaker if you're wanting to h ...
```
This is wrong. The nested resolver returning null only affects the parent resolver if the nested resolver's type in schema is defined as non-null. Since you've defined in schema that the nested resolver will never return null, when it does return null that's an error according to graphql engine. Since by the promise of schema it cannot return null for that nested resolver, the error traverses up to the parents until the engine finds a field which is defined as nullable in the schema.

By default the root `data` field is nullable in graphql, it's implemented in graphql engine core. So, null value for a nested resolver results in the whole data object being returned as null.

This situation can easily be avoided if you declare nested resolvers as nullable in schema.

```graphql
type User {
  posts: PostsConnection!
}
```

```graphql
type User {
  posts: PostsConnection
}
```

Notice the difference? The difference is `!` exclamation sign which makes it nullable/non-nullable. The concept of partial data is exactly this. If a nested resolver is nullable, you can return null or throw an error inside that resolver, and it will have no effect on the other fields. In case of throwing error the nested resolver field will have a value of null and in root errors array the error you threw in nested resolver will be present. The error will also exactly mention which field the error occured in, in its path field. Something like this:-
```ts
{
  path: ["user", "posts"]
}
```

kb-0311 replied:
@xoldyckk Oh, I had no idea that error could be avoided by making the return type nullable for nested resolver. 

Should I remove that sentence regarding nested resolver?

>By that this means even if some nested resolver fails, that has the impact on the result of the root parent resolver (contradicts Introduction point 4).Essentially failure of a related node leads to the failure of parent node.This can potentially be a deal breaker if you're wanting to have errors returned as part of a mutation, but want to query for data on the result anyways. A common example is returning an error with some partial data for private Organization(name , location etc) even if a lot of background details like members , admins etc should not be returned.

xoldd replied:
@kb-0311 If you had no idea about it then you should learn more about it and it's workings. Might change the mental model of how you're currently perceiving things, will probably let you think of more flexible approaches.

And yeah of course you need to get rid of those lines, you're giving out incorrect knowledge to the readers.

kb-0311 replied:
@xoldyckk Yeah that is the tricky part there is not much info or a universally agreed upon approach regarding Graphql Error handling other than the stage 6a one and stage 5 one and a lot of stuff regarding best practises for error handling keeps changing. 

Even if we consider the nullable point you raised  I doubt it would be enough to completely change the approach ->
>In case of throwing error the nested resolver field will have a value of null and in root errors array the error you threw in nested resolver will be present. The error will also exactly mention which field the error occured in, in its path field. 

There is the matter of Apollo client since the default graphql errors array in there specifies the path and considers that as the differentiating factor , what happens when we are destructuring `data` and `error` object from a `useQuey` hook?  

The client apps will show an error for the whole query since any error in the root error array will cause it to fail.

There are ways around this on the client apps ofc but I cannot think one which would be simpler than this one currently.


Though I am not here to discuss about the apollo client again, just that I do think right now the 6a approach is looking kinda attractive ngl. But will think about more approaches

xoldd commented on docs/developers/talawa-api/error-handling.md:
````diff
@@ -0,0 +1,625 @@
+
+```rust
+
+data: {
````
This response is incorrect. If email is taken or password is not according to constraints the returned user object should be null. The user object you're returning is to return the user you created, here no user will be created because the constraints are never fullfilled. This is not an example of partial errors.

kb-0311 replied:
@xoldyckk Yes you are right. Although that example was meant as an example of field errors primarily. Will correct that response and remove mention of partial data there. 

So basically the nested resolver nullability and partial data concept are basically the same, right? So should I mention that in the nested resolver error section ? or make a new example with another connection for that (although I doubt it would be necessary)?

kb-0311 replied:
Also regarding the scalar fields I think we should send errors within the nested resolver itself since then it will be easier to query for the clients since all errors would be in that respective resolver. What do you think?

Also, could you tell me what example nested query from talawa-api could be used in the doc here to show scalar field error handling?

xoldd replied:
@kb-0311 Think of nested resolver as just a field. The field can be scalar or a complex objects. Whether you choose to implement a seperate resolver for scalar fields or not depends on your implementation.

Since, you return most of the scalars in the parent resolver itself graphql doesn't throw out errors to resolve the fields which are asked for in schema. This is because the fields in schema and fields in database have the same name and data type.

Same is not true for custom scalar and complex object fields. Example for custom scalar is `postsCount`, there's no database field named `postCount` in database, you implement a seperate resolver for `postsCount` and return the `postsCount` according to the parent object. Example for complex object you already know, `posts` is one example. You're not restricted to resolve all scalar values in parent resolver and all complex objects in nested resolvers. Nested resolvers work for anything. 

Something as stupid as this is possible too:-
```
Query: {
  parentResolver: () => {
    return { name: "test" }
  }
},

Parent: {
  name: (parent) => {
    return parent.name.toUpperCase();
  }
}
```

The nested resolver here transforms the parent.name. This could've been done in the parent resolver itself, but this is just an example, you're free to do whatever you want with the values parent resolver returns in the nested resolvers. When graphql resolves parent.name field, it won't take the name you returned from parentResolver, it'll make use of the custom resolver you defined for parent.name. So, custom resolver take more precedence over the values you simply returned in parent resolver. If there's no custom resolver graphql will simply return the parent.name you returned in parent resolver. All it cares about is that the name and type of field you return matches with what's defined in the schema. 

Now whether you choose to implement data/errors object for each of these field(scalar or object) resolvers is up to you. This is something I've explained to you many times now.  This is one of the problems with errors in schema approach, too much boilerplate. In root response object errors list, you just throw the error in any resolver and it lands in the list.

Do keep in mind, some scalar fields are database intensive to compute. I don't need to give an example here, you can probably think of many scalar fields like these. If you implement them in parent resolver, you're computing those scalar fields each time someone makes a request. That's bad for the database and server.

Implementing things in nested resolvers is one way of making sure computationally expensive fields don't get triggered unnecessarily. There's another way, it's called `look ahead` pattern. In this you take the graphql info object, parse it and only compute the specific fields asked by the clients. But this is high level complexity, because doing this manually would be braindead. Automating this is only possible if your graphql schema and database schema can be mapped and interconverted using some kind of compiler or something. This is something only tech giants would do.

kb-0311 replied:
@xoldyckk like organizationMemberConnection and its `email` field for members be a good example to demonstrate the scalar field error in a nested resolver?

kb-0311 replied:
@xoldyckk Ohh alright got it , thanks for explaining that so clearly and in detail.

like One possible organizationMemberConnection and its email field (which can be a custom resolver/scalar) for members be a good example to demonstrate the scalar field error in a nested resolver, right?

Because in the future that field is a PII field and will be encrypted in the Database and in transit even within the SSL layer, as asked by Peter. So for that scalar, it would make a lot of sense to have it as a separate resolver for its decryption based on some form of access control.

Should I use this example if it makes sense to you?

xoldd replied:
@kb-0311 The point is just to show that nested resolvers are not something magical, they're needed for properly organizing the code and having a clean architecture, delegating computation of expensive or custom fields which are ephimeral i.e., only exist at the graphql level not the database. `email` is a good example for this, but you can use whatever you want like `bank_account_number`, `bank_balance`, `phone_number`, `postsCount` etc.

xoldd replied:
@kb-0311 When you resolve custom/complex fields within the parent resolver, you have the freedom to return errors related to those fields within the errors field of that parent resolver. It's possible to resolve your whole graphql schema and all the relations within a single resolver, but that's anti graphql. You see this is only different  from the root errors approach in that here errors will be type-safe, in root errors list they're not type-safe, you'll still be resolving errors for all fields/nested fields within the parent resolver, and that'll make a big list of possible errors that parent resolver can return.

Extracting data and errors into resolvers is a modular approach and isolates information of a field to that field's resolver, the resolvers can be for scalar or complex objects. But this does result in the graphql schema not representing the relations directly, instead you have `{success,errors}` objects for many fields(scalar or complex) which looks weird.

So, the tradeoffs are:-
1. **_non type safe errors_** -> cleaner graphql schema, harder to colocate errors to where they occured as all of them are dumped in the root errors list, no type safety, harder to handle them
2. _**type safe errors**_ -> weird/complex graphql schema, errors are colocated with the field they occured for, type safe, easier to handle them

kb-0311 replied:
Oh yeah, I agree with you completely. In talawa, in the future, we can even introduce fields that are not necessarily defined cut to cut in the database but could be valuable data to be sent back to the client. The example you stated yourself included.

I could not agree more with the tradeoff points you made. Ultimately I will also need to talk to the client app devs to discuss how the corresponding UI elements will be developed for this 6a approach too since a lot of the motive behind this was to allow more customization in errors within the apps.

xoldd reviewed (commented):
You also have not mentioned about errors related to nested resolvers that resolve a scalar field like `accountBalance`, `phoneNumber`, `email` etc., which are fields that do have RBAC associated to them. Now whether you return errors related to scalars within the parent resolver's errors list, or within their own resolver's errors list is your choice.

kb-0311 committed:
scalar fields

kb-0311 commented:
@xoldyckk 

Added the distinction of nullability in the nested resolver point.

fixed the response for sign up mutation to send multiple errors.

Added the scalar field custom resolver logic.Since some scalar fields are database intensive to compute within their own custom resolver so it should be used sparingly and because there are not a lot of fields which would require access control I think it is alright to have the errors thrown in their fields at the cost of boilerplate.

kb-0311 committed:
short note about nested resolvers

kb-0311 commented:
@xoldyckk Added the points about nested resolvers for scalar fields as a short note and the tradeoffs between type-safe and non-type-safe errors in this doc too.

Pls do review, anything else we should tweak?

kb-0311 committed:
small tweak

kb-0311 committed:
small tweak

kb-0311 committed:
small tweak

kb-0311 commented:
@xoldyckk is the doc okay now?
Can we merge this?

xoldd commented:
@kb-0311 

The doc does contain most of the necessary stuff but I think it can use some better
examples, organization and grouping of things. Also, it has many grammatical mistakes
fix them, docs are important part of software projects try to be as professional as you can
be.

Make a table of contents at the top linking to different sections/subsections to make
navigation easier.

You don't need to copy the talawa-api schema in example, use a simple schema which
includes all the possible things you want to explain regarding errors. It should
include a mutation resolver, two field resolvers, one for scalar
field and one for complex field, and two fields which don't need a resolver one for 
a scalar value and one for a complex object. You don't need query resolver because mutation
resolver can cover it. Try to emphasize that errors in schema is not only limited 
to mutation resolvers and field resolvers only. 

You don't need to explain about graphql error bubbling cuz of non-nullable fields,
just link some docs for it. It's more of a schema design choice and less related
to error handling. So, in your simple schema make use of nullable fields because
you need to show the error example in nested resolvers.

Implement resolvers, explain the constraints/logic implemented in the resolvers,
and the error throw/response returned in the resolver. I'd say for this bootstrap a local dummy graphql server
to check out how errors are returned for field resolvers. Show what the errors/response
looks like in the result, for the mutation resolver and field resolvers, why it's bad.

Explain the difference between user errors and developer errors, and why user errors
should be considered as a form of data instead of errors.

Now come to the errors in schema approach, link the blog for it and the section
you're going to implement. Modify the simple schema you wrote previously to also 
include errors in schema. Show what the error/response looks like in the result, why it's better. You can use graphiql
screenshot to show the advantages of type-safety.

Explain how to decide when to model errors in schema(both scalar and complex) for fields and when not to.
`id` field scalar if not provided should result in the whole resolver erroring out, not just the `id` field.
`bankBalance` scalar field should not result in whole resolver erroring out. `image` object field should not
result in errors if business logic dictates that a certain portion of a user's profile should be viewable by
anyone including the profile picture. `friends` list field should error out if an unknown/unauthorized user
tries to access `friends` list of a private user. Again you're free to use your own examples, if you think about
it you can come up with examples of many real world examples for these fields.

Explain the tradeoffs of both approaches, probably in a tabular way.

I believe I've given you more than enough examples now. You need to make your own effort to best represent
this information in a nice and clean hierarchical way. The flow of information should be top to bottom, where
at the top you discuss the traditional way of handling errors and at the bottom how it should be
done with errors in schema approach. So, two main sections one with default graphql error handling and one with
custom error handling. Make use of subsections to better organize the information.

palisadoes commented:
@kb-0311 What's the status on this?

kb-0311 commented:
@palisadoes I am preoccupied with the End Semester exams, projects, viva practicals, etc which will end on 10th May. After that, I have a 3-month long vacation during which I will spend most of my time contributing.

kb-0311 committed:
new doc

kb-0311 committed:
navigation

kb-0311 committed:
Update error-handling.md

kb-0311 commented:
@xoldyckk @kevoniat Pls Review.

xoldd commented on docs/developers/talawa-api/error-handling.md:
```diff
@@ -0,0 +1,1110 @@
+          // Here we will be returning multiple errors in the form on an array of signUpErrors
+          if (signUpErrors) {
+               return  {
+                   signUpData: {
```
Here make signUpData return null.

xoldd commented on docs/developers/talawa-api/error-handling.md:
```diff
@@ -0,0 +1,1110 @@
+            }
+
+          // Here we will be returning multiple errors in the form on an array of signUpErrors
+          if (signUpErrors) {
```
Empty object(array) in javascript is considered a truthy value, this would trigger the if block it even though there might be no errors. Instead of this make the check on whether the array is empty or not. Similarly, in schema make the errors list non nullable, that way an array is always returned but client has to check if it is empty or not to determine whether an error occured.

Other choice is to only return an array for errors field when it has at least one item inside it, otherwise just return null. This wouldn't require changes to the schema.

xoldd commented on docs/developers/talawa-api/error-handling.md:
```diff
@@ -0,0 +1,1110 @@
+
+data: {
+  signUpData: {
+      user: null,
```
Again this is wrong at the schema level, you've defined the fields inside signUpData to be non nullable, but you're returning null for those fields here. Instead you should return null for the signUpData field as a whole.

Make sure the schema you use and the resolvers you implement map to each other correctly. Applies for all the examples you've used in this doc.

xoldd reviewed (commented):
@kb-0311 Maybe run the doc through some grammer/spelling mistakes checker tool to fix the minor issues. Overall pretty well thought out doc. Great job.

Other things left are the design guide for schema(thinking in terms of nodes, edges, relations), comments(to help developers), documentation(fine grained documentation for each and every field within schema), when to use custom scalars(what should be the job of schema and what should be the job of resolvers), when to use nullable/non-nullable fields, notifying deprecations and schema updates/migration to clients etc.

kb-0311 committed:
done

kb-0311 committed:
done

kb-0311 commented:
@xoldyckk Nice catch, made the necessary changes. Pls, do review it again.

The schema design docs will be done next week. This week will probably implement the errors in schema approach in 5-6 PR to complete it in parts.

Also used Grammarly to improve the grammar.
